function-server/src/main/java/io/dispatchframework/javabaseimage/Server.java

```diff
@@ -0,0 +1,158 @@
+                .start();
+    }
+
+    public static class ExecFunction implements HttpHandler {
+        private String packageName;
+        private String className;
+        private final Gson gson;
+
+        public ExecFunction(String packageName, String className) {
+            this.packageName = packageName;
+            this.className = className;
+            this.gson = new GsonBuilder().serializeNulls().create();
+        }
+
+        public Type[] getBiFunctionTypes(Class<?> functionClass) {
+            Type[] genericInterfaces = functionClass.getGenericInterfaces();
+            Type[] genericTypes = null;
+
+            for (Type genericInterface : genericInterfaces) {
+                if (genericInterface instanceof ParameterizedType) {
+                    ParameterizedType parameterizedType = (ParameterizedType) genericInterface;
+                    if (parameterizedType.getRawType().equals(BiFunction.class)) {
+                        genericTypes = parameterizedType.getActualTypeArguments();
+                        break;
+                    }
+                }
+            }
+
+            return genericTypes;
+        }
+
+        public Object applyFunction(String message, BiFunction f, Type[] biFunctionTypes) {
+            JsonObject rootObj = new JsonParser().parse(message).getAsJsonObject();
+            String context = rootObj.get("context").toString();
+            String payload = rootObj.get("payload").toString();
+
+            if (biFunctionTypes != null && biFunctionTypes.length >= 2) {
+                return f.apply(gson.fromJson(context, biFunctionTypes[0]), gson.fromJson(payload, biFunctionTypes[1]));
+            } else {
+                return f.apply(null, null);
+            }
+        }
+
+        @Override
+        public void handleRequest(final HttpServerExchange exchange) {
+            exchange.getRequestReceiver().receiveFullString((httpServerExchange, message) -> {
+                Object r = null;
+                Exception err = null;
+
+                // Closing a ByteArrayOutputStream has no effect
+                ByteArrayOutputStream baosStderr = new ByteArrayOutputStream();
+                ByteArrayOutputStream baosStdout = new ByteArrayOutputStream();
+
+                PrintStream oldStderr = System.err;
+                PrintStream oldStdout = System.out;
+
+                try (PrintStream stderr = new PrintStream(baosStderr);
+                     PrintStream stdout = new PrintStream(baosStdout)) {
+                    try {
+                        System.setErr(stderr);
+                        System.setOut(stdout);
+
+                        Class<?> functionClass = Class.forName(packageName + "." + className);
+                        Type[] biFunctionTypes = getBiFunctionTypes(functionClass);
+                        BiFunction f = (BiFunction) functionClass.newInstance();
+
+                        r = applyFunction(message, f, biFunctionTypes);
```

Comment:
`biFunctionTypes` should be a private final field of ExecFunction, initialized and checked once (in the constructor). There's no need to recreate it every time the function is invoked and pass as an argument. 

I'd argue that `f` should also be a private final field, i.e. we'll have only one instance of the function per function server instance. Also, init it in the constructor. 

With this change `functionClass` var will get moved to the constructor. Also, `packageName` and `className` fields become unnecessary. 

Thus `applyFunction(message, f, biFunctionTypes)` should become `applyFunction(message)`

Reply:
Oh, and it will make the tests shorter too ;)

Reply:
For the sake of ease of testing, those fields don't necessarily need to be private. Package private (the default) would be sufficient.

Reply:
I very much agree with this change.